### PR TITLE
[IM] Fix verious corner cases for event chunking

### PR DESCRIPTION
**Pull request by erjiaqing**
#### Problem
There are some corner cases when reading events and attributes at the same time.
- What if attribute data used all buffers, and we cannot try to encode anything?
  > We need to reserve a few bytes so we can at least check if we can encode anything.
- What if we have encoded some attributes, and the event is too large to fit into the remaining space.
  > We should not skip it in this case, let's wait for next chunk.
- We may miss one event between chunks (increasing counter by mistake)
  > Do not increase aEventMin when buffer is too small to fit current event data, it should be reported in the next chunk. 
- When attributes is followed by a series of chunks of events, OnReportEnd will not be called.
  > Record a transaction wide `mIsAttributeReportIBsPresent` to deliver OnReportEnd() at the end of report. 

#### Change overview

The cases above is fixed, and is verified by the tests added.

#### Testing

src/controller/tests/TestReadChunking.cpp

### CODE REVIEW

File: src/app/ReadClient.h [hunk 1]
```diff
@@ -339,7 +339,8 @@ class ReadClient : public Messaging::ExchangeDelegate
     FabricIndex mFabricIndex            = kUndefinedFabricIndex;
     InteractionType mInteractionType    = InteractionType::Read;
     Timestamp mEventTimestamp;
-    EventNumber mEventMin = 0;
+    EventNumber mEventMin             = 0;
+    bool mIsAttributeReportIBsPresent = false;
```
Comment (bzbarsky-apple) [inline src/app/ReadClient.h:343]: Please document the semantics of this member.  It's really not clear from the name what it means....

And maybe it should be named `mSawAttributeReportsInCurrentReport` or something.

File: src/app/reporting/Engine.cpp [hunk 2]
```diff
@@ -302,9 +306,9 @@ CHIP_ERROR Engine::BuildSingleReportDataEventReports(ReportDataMessage::Builder
         else if ((err == CHIP_ERROR_BUFFER_TOO_SMALL) || (err == CHIP_ERROR_NO_MEMORY))
         {
             // when first cluster event is too big to fit in the packet, ignore that cluster event.
-            if (eventCount == 0)
+            // However, we may have encoded some attributes befure, we don't skip it in that case.
```
Comment (bzbarsky-apple) [inline src/app/reporting/Engine.cpp:309]:
```suggestion
            // However, we may have encoded some attributes before, we don't skip it in that case.
```

File: src/app/reporting/Engine.cpp [hunk 3]
```diff
@@ -302,9 +306,9 @@ CHIP_ERROR Engine::BuildSingleReportDataEventReports(ReportDataMessage::Builder
...
+            if (eventCount == 0 && !aBufferIsUsed)
             {
-                eventMin++;
```
Comment (bzbarsky-apple) [inline src/app/reporting/Engine.cpp:307]: Shouldn't this still increment eventMin?  How does that match the "ignore that cluster event" comments above?

File: src/app/reporting/Engine.cpp [hunk 4]
```diff
@@ -289,8 +290,11 @@ CHIP_ERROR Engine::BuildSingleReportDataEventReports(ReportDataMessage::Builder
     }
 
     {
-        EventReportIBs::Builder & eventReportIBs = aReportDataBuilder.CreateEventReports();
+        const uint32_t kReservedSizeEndOfReportIBs = 1;
```
Comment (bzbarsky-apple) [inline src/app/reporting/Engine.cpp:293]: So we're reserving 1 byte?  Might be worth clearly documenting why one byte (for the TLV end-of-struct marker, I am assuming?).

File: src/app/reporting/Engine.cpp [hunk 5]
```diff
@@ -338,9 +343,13 @@ CHIP_ERROR Engine::BuildSingleReportDataEventReports(ReportDataMessage::Builder
         *apHasEncodedData = !(eventCount == 0 || eventClean);
     }
 
-    if (err == CHIP_NO_ERROR && (eventCount == 0 || eventClean))
+    // Maybe encoding the attributes have already used up all space.
```
Comment (bzbarsky-apple) [inline src/app/reporting/Engine.cpp:346]:
```suggestion
    // Maybe encoding the attributes has already used up all space.
```

File: src/app/reporting/Engine.cpp (hunk 4 above)
Comment (erjiaqing) [reply src/app/reporting/Engine.cpp:295]: Added a comment for it

File: src/app/reporting/Engine.cpp (hunk 3 above)
Comment (erjiaqing) [reply src/app/reporting/Engine.cpp:307]: Yes, this is removed by mistake, eventMin should be increased by 1 when the buffer is clean before encoding.

File: src/app/ReadClient.h (hunk 1 above)
Comment (erjiaqing) [reply src/app/ReadClient.h]: mSawAttributeReportsInCurrentReport is a better name and is self-explained.

File: src/app/reporting/Engine.cpp [hunk 6]
```diff
@@ -302,9 +308,13 @@ CHIP_ERROR Engine::BuildSingleReportDataEventReports(ReportDataMessage::Builder
         else if ((err == CHIP_ERROR_BUFFER_TOO_SMALL) || (err == CHIP_ERROR_NO_MEMORY))
         {
             // when first cluster event is too big to fit in the packet, ignore that cluster event.
+            // However, we may have encoded some attributes befoure, we don't skip it in that case.
```
Comment (bzbarsky-apple) [inline src/app/reporting/Engine.cpp:311]:
```suggestion
            // However, we may have encoded some attributes before, we don't skip it in that case.
```

**Review by bzbarsky-apple — APPROVED** [review]

**Review by yunhanw-google — APPROVED** [review]

**Pull request merged**